### PR TITLE
`[Move]` Part-4 Classes into Different Library - `Neo.Extensions`

### PR DESCRIPTION
**Pull request by cschuchardt88**
# Description
Reorganized and move classes from `neo` library to `neo.extensions` library. 

### Fixed problems and warnings that occurred. Along with naming violations.

# Change Log
 - Added `HashSetExtensions`, `HashSetExtensions2`, `IpAddressExtensions`, `AssemblyExtensions`, `StringExtensdions` classes
 - Deleted `./src/Neo/Helper.cs` _(no more methods in file)_

## Type of change

- [x] Optimization (the change is only an optimization)
- [x] Style (the change is only a code style for better maintenance or standard purpose)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
- Unit Testing

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by Jim8y** [comment]
Please resolve conflicts.

**Comment by cschuchardt88** [comment]
> Please resolve conflicts.

image

But done

**Comment by Jim8y** [comment]
@cschuchardt88 conflicts

File: src/Neo/Extensions/HashSetExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+using Neo.IO.Caching;
+using System;
+using System.Collections.Generic;
+
+namespace Neo.Extensions
+{
+    /// <summary>
+    /// A helper class that provides common functions.
+    /// </summary>
+    public static class HashSetExtensions2
```
Comment (Jim8y) [inline src/Neo/Extensions/HashSetExtensions.cs:21]: extension2?

Comment (cschuchardt88) [reply src/Neo/Extensions/HashSetExtensions.cs:21]: yes we already have `HashSetExtensions` in the `neo.extensions.dll`

File: src/Neo.Extensions/RandomExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+using System;
+using System.Numerics;
+
+namespace Neo.Extensions
+{
+    public static class RandomExtensions
+    {
+        public static BigInteger NextBigInteger(this Random rand, int sizeInBits)
+        {
+            if (sizeInBits < 0)
```
Comment (Jim8y) [inline src/Neo.Extensions/RandomExtensions.cs:21]: quick question, if it must be positive, why not ushort

Comment (cschuchardt88) [reply src/Neo.Extensions/RandomExtensions.cs:21]: I'm just moving the function to extensions library.

**Review by Jim8y — APPROVED** [review]

File: src/Neo/Extensions/HashSetExtensions.cs (hunk 1 above)
Comment (shargon) [reply src/Neo/Extensions/HashSetExtensions.cs:21]: Why not join them?

**Comment by shargon** [comment]
Wait for the naming issue

File: src/Neo/Extensions/HashSetExtensions.cs (hunk 1 above)
Comment (cschuchardt88) [reply src/Neo/Extensions/HashSetExtensions.cs:21]: `HashSetExtensions` doesnt require any dependencies inside `neo` library and `HashSetExtensions2` does. So can't be done

Comment (shargon) [reply src/Neo/Extensions/HashSetExtensions.cs:21]: We can use the same name with different namespaces, but the 2, it's weird

Comment (cschuchardt88) [reply src/Neo/Extensions/HashSetExtensions.cs:21]: If you insist, but you won't see this name and it shouldn't be used.

File: tests/Neo.Extensions.Tests/UT_BigIntegerExtensions.cs [hunk 3]
```diff
@@ -18,6 +18,7 @@ namespace Neo.Extensions.Tests
     [TestClass]
     public class UT_BigIntegerExtensions
     {
+
         [TestMethod]
```
Comment (shargon) [inline tests/Neo.Extensions.Tests/UT_BigIntegerExtensions.cs:22]:
```suggestion
        [TestMethod]
```

File: src/Neo.Extensions/StringExtensdions.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+//
+// Redistribution and use in source and binary forms with or without
+// modifications are permitted.
+
+using System;
+using System.Globalization;
+
+namespace Neo.Extensions
+{
+    public static class StringExtensdions
```
Comment (shargon) [inline src/Neo.Extensions/StringExtensdions.cs]:
```suggestion
    public static class StringExtensions
```

**Review by shargon — DISMISSED** [review]

**Comment by cschuchardt88** [comment]
@Jim8y @shargon had to run `dotnet format`

**Review by Jim8y — APPROVED** [review]

**Comment by Jim8y** [comment]
@shargon please review again

**Review by AnnaShaleva — APPROVED** [review]

**Pull request merged**